File: AuctionMarket/src/main/java/com/portfolio/auctionmarket/domain/products/repository/ProductQueryRepository.java
package com.portfolio.auctionmarket.domain.products.repository;


import com.portfolio.auctionmarket.domain.auctions.entity.AuctionStatus;
import com.portfolio.auctionmarket.domain.products.dto.ProductListCondition;
import com.portfolio.auctionmarket.domain.products.entity.Product;
import com.portfolio.auctionmarket.domain.products.entity.ProductStatus;
import com.querydsl.core.types.OrderSpecifier;
import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQuery;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.domain.SliceImpl;
import org.springframework.data.support.PageableExecutionUtils;
import org.springframework.stereotype.Repository;
import org.springframework.util.StringUtils;

import java.util.List;

import static com.portfolio.auctionmarket.domain.auctions.entity.QAuction.auction;
import static com.portfolio.auctionmarket.domain.bids.entity.QBid.bid;
import static com.portfolio.auctionmarket.domain.products.entity.QProduct.product;
import static com.portfolio.auctionmarket.domain.categories.entity.QCategory.category1;
import static com.portfolio.auctionmarket.domain.user.entity.QUser.user;

@Repository
@RequiredArgsConstructor
public class ProductQueryRepository {

    private final JPAQueryFactory jpaQueryFactory;;

    public Page<Product> productList(Long userId, ProductListCondition condition, Pageable pageable) {
        List<Product> content = jpaQueryFactory
                .selectFrom(product)
                .innerJoin(product.seller, user).fetchJoin()
                .innerJoin(product.auction, auction).fetchJoin()
                .leftJoin(product.category, category1).fetchJoin()
                .where(
                        isMyAuction(userId),
                        titleContain(condition.title()),
                        pathStartWith(condition.path()),
                        statusFilter(userId),
                        auction.status.eq(AuctionStatus.PROCEEDING),
                        product.productStatus.eq(ProductStatus.ACTIVE)
                )
                .offset(pageable.getOffset())
                .limit(pageable.getPageSize())
                .orderBy(orderSpecifier(condition.sort()))
                .fetch();

        JPAQuery<Long> count = jpaQueryFactory
                .select(product.count())
                .from(product)
                .innerJoin(product.seller, user)
                .innerJoin(product.auction, auction)
                .leftJoin(product.category, category1)
                .where(
                        isMyAuction(userId),
                        titleContain(condition.title()),
                        pathStartWith(condition.path()),
                        statusFilter(userId),
                        auction.status.eq(AuctionStatus.PROCEEDING),
                        product.productStatus.eq(ProductStatus.ACTIVE)
                );

        return PageableExecutionUtils.getPage(content, pageable, count::fetchOne);
    }

    public Slice<Product> adminProductList(Long userId, ProductListCondition condition, Pageable pageable) {

        int pageSize = pageable.getPageSize();

        List<Product> content = jpaQueryFactory
                .selectFrom(product)
                .innerJoin(product.seller, user).fetchJoin()
                .innerJoin(product.auction, auction).fetchJoin()
                .leftJoin(product.category, category1).fetchJoin()
                .where(
                        isMyAuction(userId),
                        titleContain(condition.title()),
                        pathStartWith(condition.path()),
                        statusFilter(userId)
                )
                .offset(pageable.getOffset())
                .limit(pageSize + 1)
                .orderBy(orderSpecifier(condition.sort()))
                .fetch();

        boolean hasNext = false;
        if (content.size() > pageSize) {
            content.remove(pageSize);
            hasNext = true;
        }

        return new SliceImpl<>(content, pageable, hasNext);
    }

    private BooleanExpression isMyAuction(Long userId) {
        return userId != null ? product.seller.userId.eq(userId) : null;
    }
    private BooleanExpression titleContain(String title) {
        return StringUtils.hasText(title) ? product.title.contains(title) : null;
    }
    private BooleanExpression pathStartWith(String path) {
        if (!StringUtils.hasText(path)) {
            return null;
        }

        return product.category.path.eq(path)
                .or(product.category.path.startsWith(path + "/"));
    }

    private BooleanExpression statusFilter(Long userId) {
        if (userId == null) {
            return null;
        }
        return product.productStatus.in(ProductStatus.ACTIVE, ProductStatus.SOLD, ProductStatus.FAILED);
    }

    private OrderSpecifier<?> orderSpecifier(String sort) {
        if ("endingSoon".equals(sort)) {
            return product.auction.endTime.asc();
        } else if ("priceHigh".equals(sort)) {
            return product.auction.currentPrice.desc();
        } else if ("priceLow".equals(sort)) {
            return product.auction.currentPrice.asc();
        }
        return product.productId.desc();
    }
}

File: AuctionMarket/auction-domain/src/main/kotlin/org/example/auctioncommon/domain/product/repository/ProductQueryRepository.kt
package org.example.auctioncommon.domain.product.repository

import com.querydsl.core.types.OrderSpecifier
import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQuery
import com.querydsl.jpa.impl.JPAQueryFactory
import org.example.auctioncommon.domain.auction.entity.AuctionStatus
import org.example.auctioncommon.domain.auction.entity.QAuction.auction
import org.example.auctioncommon.domain.categories.entity.QCategory.category1
import org.example.auctioncommon.domain.product.dto.ProductListCondition
import org.example.auctioncommon.domain.product.entity.Product
import org.example.auctioncommon.domain.product.entity.ProductStatus
import org.example.auctioncommon.domain.product.entity.QProduct.product
import org.example.auctioncommon.domain.user.entity.QUser.user
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Slice
import org.springframework.data.domain.SliceImpl
import org.springframework.data.support.PageableExecutionUtils
import org.springframework.stereotype.Repository
import java.util.function.LongSupplier

@Repository
class ProductQueryRepository(
    private val jpaQueryFactory: JPAQueryFactory
) {

    fun productList(
        userId: Long?,
        condition: ProductListCondition,
        pageable: Pageable
    ): Page<Product> {
        val content = jpaQueryFactory
            .selectFrom(product)
            .innerJoin(product.seller, user).fetchJoin()
            .innerJoin(product.auction, auction).fetchJoin()
            .leftJoin(product.category, category1).fetchJoin()
            .where(
                isMyAuction(userId),
                titleContain(condition.title),
                pathStartWith(condition.path),
                statusFilter(userId),
                auction.status.eq(AuctionStatus.PROCEEDING),
                product.productStatus.eq(ProductStatus.ACTIVE)
            )
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .orderBy(orderSpecifier(condition.sort))
            .fetch()

        val count: JPAQuery<Long?> = jpaQueryFactory
            .select(product.count())
            .from(product)
            .innerJoin(product.seller, user)
            .innerJoin(product.auction, auction)
            .leftJoin(product.category, category1)
            .where(
                isMyAuction(userId),
                titleContain(condition.title),
                pathStartWith(condition.path),
                statusFilter(userId),
                auction.status.eq(AuctionStatus.PROCEEDING),
                product.productStatus.eq(ProductStatus.ACTIVE)
            )

        return PageableExecutionUtils.getPage<Product>(content, pageable, LongSupplier { count.fetchOne() ?: 0L })
    }

    fun adminProductList(
        userId: Long?,
        condition: ProductListCondition,
        pageable: Pageable
    ): Slice<Product> {
        val pageSize = pageable.pageSize

        val content = jpaQueryFactory
            .selectFrom(product)
            .innerJoin(product.seller, user).fetchJoin()
            .innerJoin(product.auction, auction).fetchJoin()
            .leftJoin(product.category, category1).fetchJoin()
            .where(
                isMyAuction(userId),
                titleContain(condition.title),
                pathStartWith(condition.path),
                statusFilter(userId)
            )
            .offset(pageable.offset)
            .limit(pageSize.toLong() + 1)
            .orderBy(orderSpecifier(condition.sort))
            .fetch()

        var hasNext = false
        if (content.size > pageSize) {
            content.removeAt(pageSize)
            hasNext = true
        }

        return SliceImpl(content, pageable, hasNext)
    }

    private fun isMyAuction(userId: Long?): BooleanExpression? {
        return userId.let { product.seller.userId.eq(it) }
    }

    private fun titleContain(title: String?): BooleanExpression? {
        return if (!title.isNullOrBlank()) product.title.contains(title) else null
    }

    private fun pathStartWith(path: String?): BooleanExpression? {
        return if (!path.isNullOrBlank()) product.category.path.eq(path)
            .or(product.category.path.startsWith("$path/")) else null
    }

    private fun statusFilter(userId: Long?): BooleanExpression? {
        return if (userId != null) {
            product.productStatus.`in`(ProductStatus.ACTIVE, ProductStatus.SOLD, ProductStatus.FAILED)
        } else null
    }

    private fun orderSpecifier(sort: String?): OrderSpecifier<*> {
        return when (sort) {
            "endingSoon" -> product.auction.endTime.asc()
            "priceHigh" -> product.auction.currentPrice.desc()
            "priceLow" -> product.auction.currentPrice.asc()
            else -> product.productId.desc()
            }
    }
}
